streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -103,62 +103,67 @@ public void validatePartitionExists(final TopicPartition topicPartition, final S
     public void register(final StateRestorer restorer) {
         restorer.setGlobalRestoreListener(stateRestoreListener);
         stateRestorers.put(restorer.partition(), restorer);
+        needsInitializing.put(restorer.partition(), restorer);
     }
 
 
     public Collection<TopicPartition> restore() {
-        final List<TopicPartition> completed = new ArrayList<>();
-        if (!initialized) {
-            completed.addAll(initialize());
+        if (!needsInitializing.isEmpty()) {
+            initialize();
         }
 
         if (needsRestoring.isEmpty()) {
             consumer.assign(Collections.<TopicPartition>emptyList());
-            return completed;
+            return completed();
         }
 
         final Set<TopicPartition> partitions = new HashSet<>(needsRestoring.keySet());
         final ConsumerRecords<byte[], byte[]> allRecords = consumer.poll(10);
         for (final TopicPartition partition : partitions) {
-            if (restorePartition(allRecords, partition)) {
-                completed.add(partition);
-            }
+            restorePartition(allRecords, partition);
         }
 
         if (needsRestoring.isEmpty()) {
             consumer.assign(Collections.<TopicPartition>emptyList());
         }
-        return completed;
+
+        return completed();
     }
 
-    private Collection<TopicPartition> initialize() {
-        needsRestoring.clear();
+    private void initialize() {
+        final Map<TopicPartition, StateRestorer> newTasksNeedingRestoration = new HashMap<>();
+
         if (!consumer.subscription().isEmpty()) {
             throw new IllegalStateException(String.format("Restore consumer should have not subscribed to any partitions (%s) beforehand", consumer.subscription()));
         }
-        endOffsets.putAll(consumer.endOffsets(stateRestorers.keySet()));
+        endOffsets.putAll(consumer.endOffsets(needsInitializing.keySet()));
 
         // remove any partitions where we already have all of the data
         for (final Map.Entry<TopicPartition, Long> entry : endOffsets.entrySet()) {
             TopicPartition topicPartition = entry.getKey();
             Long offset = entry.getValue();
-            final StateRestorer restorer = stateRestorers.get(topicPartition);
-            if (restorer.checkpoint() >= offset) {
-                restorer.setRestoredOffset(restorer.checkpoint());
-            } else if (restorer.offsetLimit() == 0) {
-                restorer.setRestoredOffset(0);
-            } else {
-                needsRestoring.put(topicPartition, restorer);
-                final Long endOffset = endOffsets.get(topicPartition);
-                restorer.restoreStarted(restorer.startingOffset(), endOffset);
+            final StateRestorer restorer = needsInitializing.get(topicPartition);
+            // might be null as has was initialized in a previous invocation.
```

Comment:
This comment seems still valid: we should not expect `restorer` to ever be null, right?

Reply:
yeah it could be as the partition is taken from the `endOffsets` map that has all of the partitions that are being restored. In this round it might not need initializing

Reply:
Ack.